extensions/amp-drilldown/0.1/amp-drilldown.js

```diff
@@ -0,0 +1,232 @@
+    this.document_ = this.win.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private {Array} */
+    this.historyIds_ = [];
+
+    /** @private {?Element} */
+    this.currentSubmenu_ = null;
+
+    /** @private {function(!Event)} */
+    this.submenuOpenHandler_ = this.handleSubmenuOpenClick_.bind(this);
+
+    /** @private {function(!Event)} */
+    this.submenuCloseHandler_ = this.handleSubmenuCloseClick_.bind(this);
+  }
+
+  /** @override */
+  buildCallback() {
+    userAssert(isExperimentOn(this.win, 'amp-sidebar-v2'), 'Experiment is off');
+
+    this.documentElement_.addEventListener('keydown', e =>
+      this.handleKeyDown_(e)
+    );
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.registerEventListeners_();
+    return Promise.resolve();
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.unregisterEventListeners_();
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FILL;
+  }
+
+  /**
+   * Add event listeners on submenu open/close elements.
+   */
+  registerEventListeners_() {
+    const openElements = this.element.querySelectorAll(
+      '[amp-drilldown-submenu-open]'
+    );
+    openElements.forEach(element => {
+      element.addEventListener('click', this.submenuOpenHandler_);
+    });
+    const closeElements = this.element.querySelectorAll(
+      '[amp-drilldown-submenu-close]'
+    );
+    closeElements.forEach(element => {
+      element.addEventListener('click', this.submenuCloseHandler_);
+    });
+  }
+
+  /**
+   * Remove listeners on all submenu open/close elements.
+   */
+  unregisterEventListeners_() {
+    const openElements = this.element.querySelectorAll(
+      '[amp-drilldown-submenu-open]'
+    );
+    openElements.forEach(element => {
+      element.removeEventListener('click', this.submenuOpenHandler_);
+    });
+    const closeElements = this.element.querySelectorAll(
+      '[amp-drilldown-submenu-close]'
+    );
+    closeElements.forEach(element => {
+      element.removeEventListener('click', this.submenuCloseHandler_);
+    });
+  }
+
+  /**
+   * Handler for submenu open element click.
+   * @param {Event} e
+   */
+  handleSubmenuOpenClick_(e) {
+    const submenuOpen = dev().assertElement(e.target);
+    const submenu = submenuOpen.parentElement.querySelector(
+      '[amp-drilldown-submenu]'
+    );
+    if (submenu) {
+      this.open_(submenu);
+    }
+  }
+
+  /**
+   * Open the specified submenu.
+   * @param {!Element} submenu
+   */
+  open_(submenu) {
+    if (submenu.hasAttribute('open')) {
+      return;
+    }
+    const submenuParent = this.getParentMenu_(submenu);
+    if (submenuParent) {
+      submenuParent.setAttribute('child-open', '');
+      submenu.setAttribute('open', '');
+      this.getHistory_()
+        .push(() => this.close_(submenu))
+        .then(historyId => {
+          this.historyIds_.push(historyId);
```

Comment:
Does not have to be in this PR, but make sure that we: 
1. Have history tests. 
2. Test the history interaction with amp-stories, or make sure that we restrict this component from being used in `amp-stories` via Validation.